File: csharp/solid/InterfaceSegregation.cs
namespace Solid
{
    ///
    /// <summary>
    /// Class <c>InterfaceSegregation</c> represents interface segregation principle usecase.
    /// 1. specific interfaces are better than general purpose ones
    /// 2. classes should not implement methods they don't need
    /// </summary>
    ///
    class InterfaceSegregation
    {
        public static void Run()
        {
            Console.WriteLine("\nInterface Segregation\n");

            Document doc = new Document("File");

            // general purpose interface
            MultiFunctionDocumentPrinter mfdp = new MultiFunctionDocumentPrinter();
            mfdp.Print(doc);
            mfdp.Scan(doc);

            StandardPrinter sp = new StandardPrinter();
            sp.Print(doc);
            // sp.Scan(doc); // will fail

            // specific interface
            DocumentPrinter dp = new DocumentPrinter();
            dp.Print(doc);

            MultiFunctionDocumentPrinterEnhanced mfdpe = new MultiFunctionDocumentPrinterEnhanced();
            mfdpe.Print(doc);
            mfdpe.Scan(doc);

            // utilize delegation
            DocumentScanner ds = new DocumentScanner();
            DocumentFax df = new DocumentFax();

            MultiFunctionDocumentPrinterDelegated mfdpd = new MultiFunctionDocumentPrinterDelegated(dp, ds, df);
            mfdpd.Print(doc);
            mfdpd.Scan(doc);
        }
    }

    /// <summary>Class <c>Document</c> document entity.</summary>
    class Document
    {
        public string Name;

        public Document(string name)
        {
            Name = name ?? throw new ArgumentNullException(paramName: nameof(name));
        }
    }

    /// <summary>Interface <c>IMachine</c> represents general purpose interface.</summary>
    interface IMachine<T>
    {
        void Print(T t);
        void Scan(T t);
        void Fax(T t);
    }

    /// <summary>Class <c>MultiFunctionDocumentPrinter</c> capable to perform multiple operations on documents.</summary>
    class MultiFunctionDocumentPrinter : IMachine<Document>
    {
        public void Print(Document d)
        {
            Console.WriteLine($"Print: {d.Name}");
        }
        public void Scan(Document d)
        {
            Console.WriteLine($"Scan: {d.Name}");
        }
        public void Fax(Document d)
        {
            Console.WriteLine($"Fax: {d.Name}");
        }
    }

    /// <summary>Class <c>StandardPrinter</c> capable to perform printing operations on documents.</summary>
    class StandardPrinter : IMachine<Document>
    {
        public void Print(Document d)
        {
            Console.WriteLine($"Printed: {d.Name}");
        }

        // StandardPrinter have to implement methods that useless for its purpose in order to implement IMachine interface.
        public void Scan(Document d)
        {
            throw new NotImplementedException();
        }
        public void Fax(Document d)
        {
            throw new NotImplementedException();
        }
    }

    /// <summary>Interface <c>IPrinter</c> represents specific interface.</summary>
    interface IPrinter<T>
    {
        void Print(T t);
    }

    /// <summary>Interface <c>IScanner</c> represents specific interface.</summary>
    interface IScanner<T>
    {
        void Scan(T t);
    }

    /// <summary>Interface <c>IFax</c> represents specific interface.</summary>
    interface IFax<T>
    {
        void Fax(T t);
    }

    /// <summary>Interface <c>IMultiFunctionPrinter</c> represents combinator interface.</summary>
    interface IMultiFunctionPrinter<T> : IPrinter<T>, IScanner<T>, IFax<T> { }

    /// <summary>Class <c>DocumentPrinter</c> capable to perform printing operations on documents.</summary>
    class DocumentPrinter : IPrinter<Document>
    {
        public void Print(Document d)
        {
            Console.WriteLine($"Print: {d.Name}");
        }
    }

    /// <summary>Class <c>DocumentScanner</c> capable to perform scanning operations on documents.</summary>
    class DocumentScanner : IScanner<Document>
    {
        public void Scan(Document d)
        {
            Console.WriteLine($"Scan: {d.Name}");
        }
    }

    /// <summary>Class <c>DocumentFax</c> capable to perform fax operations on documents.</summary>
    class DocumentFax : IFax<Document>
    {
        public void Fax(Document d)
        {
            Console.WriteLine($"Fax: {d.Name}");
        }
    }

    /// <summary>Class <c>MultiFunctionDocumentPrinterEnhanced</c> capable to perform multiple operations on documents.</summary>
    class MultiFunctionDocumentPrinterEnhanced : IMultiFunctionPrinter<Document>
    {
        public void Print(Document d)
        {
            Console.WriteLine($"Print: {d.Name}");
        }
        public void Scan(Document d)
        {
            Console.WriteLine($"Scan: {d.Name}");
        }
        public void Fax(Document d)
        {
            Console.WriteLine($"Fax: {d.Name}");
        }
    }

    /// <summary>Class <c>MultiFunctionDocumentPrinterDelegated</c> capable to delegate certain functionalities to other classes (Decorator).</summary>
    class MultiFunctionDocumentPrinterDelegated : IMultiFunctionPrinter<Document>
    {
        private DocumentPrinter printer;
        private DocumentScanner scanner;
        private DocumentFax fax;

        public MultiFunctionDocumentPrinterDelegated(DocumentPrinter printer, DocumentScanner scanner, DocumentFax fax)
        {
            this.printer = printer ?? throw new ArgumentNullException(paramName: nameof(printer));
            this.scanner = scanner ?? throw new ArgumentNullException(paramName: nameof(scanner));
            this.fax = fax ?? throw new ArgumentNullException(paramName: nameof(fax));
        }

        public void Print(Document d)
        {
            printer.Print(d);
        }

        public void Scan(Document d)
        {
            scanner.Scan(d);
        }

        public void Fax(Document d)
        {
            fax.Fax(d);
        }
    }
}

File: golang/solid/interface_segregation.go
package solid

import "fmt"

// I - Interface Segregation
// 1. specific interfaces are better than general purpose ones
// 2. types should not implement methods they don't need

func InterfaceSegregation() {
	fmt.Println("Interface segregation")
	fmt.Println("START")

	// doc
	d := Document{"Document1"}

	// multi func printer
	mfdp := &MultiFunctionDocumentPrinter{}
	mfdp.Print(d)

	// standard printer
	sp := &StandardPrinter{}
	sp.Print(d)

	// ---

	// printer
	pr := NewPrinter()
	pr.PrintDocument(d)

	// scanner
	sc := NewScanner()
	sc.ScanDocument(d)

	// fax
	fx := NewFax()
	fx.FaxDocument(d)

	// multi func printer enhanced
	mfpe := NewMultiFunctionDocumentPrinterEnhanced(pr, sc, fx)
	mfpe.PrintDocument(d)
	mfpe.ScanDocument(d)
	mfpe.FaxDocument(d)

	fmt.Println("END")
}

// Document - represents document.
type Document struct {
	Name string
}

// DocumentWorker - represents document worker interface.
type DocumentWorker interface {
	Print(d Document)
	Scan(d Document)
	Fax(d Document)
}

// MultiFunctionDocumentPrinter - represents multifunctional printer.
type MultiFunctionDocumentPrinter struct{}

// Print - prints document.
func (m MultiFunctionDocumentPrinter) Print(d Document) {
	fmt.Printf("Printing: %s\n", d.Name)
}

// Scan - scans document.
func (m MultiFunctionDocumentPrinter) Scan(d Document) {
	fmt.Printf("Scanning: %s\n", d.Name)
}

// Fax - sending fax.
func (m MultiFunctionDocumentPrinter) Fax(d Document) {
	fmt.Printf("Faxing: %s\n", d.Name)
}

// ---

// StandardPrinter - represents standard printer.
type StandardPrinter struct{}

// Print - prints document.
func (s StandardPrinter) Print(d Document) {
	fmt.Printf("Printing: %s\n", d.Name)
}

// Scan - scans document.
func (s StandardPrinter) Scan(_ Document) {
	// don't have functionality to scan
	panic("not supported")
}

// Fax - sending fax.
func (s StandardPrinter) Fax(_ Document) {
	// don't have functionality to fax
	panic("not supported")
}

// ---

// DocumentPrinter - represents document printer interface.
type DocumentPrinter interface {
	PrintDocument(d Document)
}

// DocumentScanner - represents document scanner interface.
type DocumentScanner interface {
	ScanDocument(d Document)
}

// DocumentFax - represents document fax interface.
type DocumentFax interface {
	FaxDocument(d Document)
}

// DocumentWorkerEnhanced - represents multifunctional document worker.
type DocumentWorkerEnhanced interface {
	DocumentPrinter
	DocumentScanner
	DocumentFax
}

// --

// Printer - represents document printer.
type Printer struct{}

// NewPrinter - creates new instance of Printer.
func NewPrinter() *Printer {
	return &Printer{}
}

// PrintDocument - prints Document.
func (s *Printer) PrintDocument(d Document) {
	fmt.Printf("Printing: %s\n", d.Name)
}

// Scanner - represents document scanner.
type Scanner struct{}

// NewScanner - creates new instance of Scanner.
func NewScanner() *Scanner {
	return &Scanner{}
}

// ScanDocument - scans Document.
func (s *Scanner) ScanDocument(d Document) {
	fmt.Printf("Scanning: %s\n", d.Name)
}

// Fax - represents document fax.
type Fax struct{}

// NewFax - creates new instance of Fax
func NewFax() *Fax {
	return &Fax{}
}

// FaxDocument - faxes Document.
func (s *Fax) FaxDocument(d Document) {
	fmt.Printf("Faxing: %s\n", d.Name)
}

// MultiFunctionDocumentPrinterEnhanced - represents multifunctional printer.
// Leverage decorator pattern.
type MultiFunctionDocumentPrinterEnhanced struct {
	p DocumentPrinter
	s DocumentScanner
	f DocumentFax
}

// NewMultiFunctionDocumentPrinterEnhanced - creates new instance of MultiFunctionDocumentPrinterEnhanced.
func NewMultiFunctionDocumentPrinterEnhanced(
	p DocumentPrinter,
	s DocumentScanner,
	f DocumentFax,
) *MultiFunctionDocumentPrinterEnhanced {
	return &MultiFunctionDocumentPrinterEnhanced{p, s, f}
}

// PrintDocument - prints Document.
func (m *MultiFunctionDocumentPrinterEnhanced) PrintDocument(d Document) {
	m.p.PrintDocument(d)
}

// ScanDocument - scans Document.
func (m *MultiFunctionDocumentPrinterEnhanced) ScanDocument(d Document) {
	m.s.ScanDocument(d)
}

// FaxDocument - faxes Document.
func (m *MultiFunctionDocumentPrinterEnhanced) FaxDocument(d Document) {
	m.f.FaxDocument(d)
}
